File: android/src/main/java/com/rtmpstreamer/livestream/VideoLiveStreamModule.kt
package com.rtmpstreamer.livestream

import android.util.Log
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.UiThreadUtil
import com.facebook.react.module.annotations.ReactModule
import com.facebook.react.uimanager.UIManagerHelper

@ReactModule(name = VideoLiveStreamModule.NAME)
class VideoLiveStreamModule(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

  override fun getName(): String = NAME

  @ReactMethod
  fun startStreaming(viewTag: Double, streamKey: String, url: String?, camera: String?) {
    withView(viewTag) { view ->
      view.startStreaming(streamKey, url, camera)
    }
  }

  @ReactMethod
  fun stopStreaming(viewTag: Double) {
    withView(viewTag) { view ->
      view.stopStreaming()
    }
  }

  @ReactMethod
  fun setZoomRatio(viewTag: Double, zoomRatio: Double) {
    withView(viewTag) { view ->
      view.setZoomRatio(zoomRatio)
    }
  }

  @ReactMethod
  fun getStreamCapabilities(viewTag: Double, camera: String?, promise: Promise) {
    UiThreadUtil.runOnUiThread {
      val resolvedViewTag = viewTag.toInt()
      val uiManager = UIManagerHelper.getUIManagerForReactTag(reactApplicationContext, resolvedViewTag)
      val view = uiManager?.resolveView(resolvedViewTag) as? VideoLiveStreamView

      if (view == null) {
        promise.reject(NAME, "Unable to resolve VideoLiveStreamView for tag $resolvedViewTag")
        return@runOnUiThread
      }

      promise.resolve(view.getStreamCapabilities(camera))
    }
  }

  private fun withView(viewTag: Double, action: (VideoLiveStreamView) -> Unit) {
    UiThreadUtil.runOnUiThread {
      val resolvedViewTag = viewTag.toInt()
      val uiManager = UIManagerHelper.getUIManagerForReactTag(reactApplicationContext, resolvedViewTag)
      val view = uiManager?.resolveView(resolvedViewTag) as? VideoLiveStreamView

      if (view == null) {
        Log.w(NAME, "Unable to resolve VideoLiveStreamView for tag $resolvedViewTag")
        return@runOnUiThread
      }

      action(view)
    }
  }

  companion object {
    const val NAME = "VideoLiveStreamModule"
  }
}

File: ios/VideoLiveStreamModule.swift
import React
import UIKit

@objc(VideoLiveStreamModule)
final class VideoLiveStreamModule: NSObject {
  @objc weak var bridge: RCTBridge?
  @objc var methodQueue: DispatchQueue {
    return bridge?.uiManager.methodQueue ?? DispatchQueue.main
  }

  @objc
  static func requiresMainQueueSetup() -> Bool {
    true
  }

  @objc
  func startStreaming(_ viewTag: NSNumber, streamKey: NSString, url: NSString?, camera: NSString?) {
    let streamKeyValue = streamKey as String
    let urlValue = url as String?
    let cameraValue = camera as String?

    withView(viewTag) { view in
      view.camera = cameraValue as NSString?
      view.startStreaming(streamKey: streamKeyValue, baseUrl: urlValue)
    }
  }

  @objc
  func stopStreaming(_ viewTag: NSNumber) {
    withView(viewTag) { view in
      view.stopStreaming()
    }
  }

  @objc
  func setZoomRatio(_ viewTag: NSNumber, zoomRatio: NSNumber) {
    withView(viewTag) { view in
      view.setZoomRatio(CGFloat(zoomRatio.doubleValue))
    }
  }

  @objc
  func getStreamCapabilities(
    _ viewTag: NSNumber,
    camera: NSString?,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    let cameraValue = camera as String?
    guard let uiManager = bridge?.uiManager else {
      reject("VideoLiveStreamModule", "bridge/uiManager not available", nil)
      return
    }

    uiManager.addUIBlock { _, viewRegistry in
      guard let view = Self.resolveView(viewTag, uiManager: uiManager, viewRegistry: viewRegistry) else {
        reject(
          "VideoLiveStreamModule",
          "Unable to find VideoLiveStreamView for reactTag \(viewTag)",
          nil
        )
        return
      }

      resolve(view.getStreamCapabilities(camera: cameraValue))
    }
  }

  private func withView(
    _ viewTag: NSNumber,
    action: @MainActor @escaping (VideoLiveStreamView) -> Void
  ) {
    guard let uiManager = bridge?.uiManager else {
      NSLog("bridge/uiManager not available")
      return
    }

    uiManager.addUIBlock { _, viewRegistry in
      guard let view = Self.resolveView(viewTag, uiManager: uiManager, viewRegistry: viewRegistry) else {
        NSLog("Unable to find VideoLiveStreamView for reactTag \(viewTag)")
        return
      }

      Task { @MainActor in
        action(view)
      }
    }
  }

  private static func resolveView(
    _ viewTag: NSNumber,
    uiManager: RCTUIManager,
    viewRegistry: [NSNumber: UIView]?
  ) -> VideoLiveStreamView? {
    if let directView = viewRegistry?[viewTag] {
      let resolvedView = RCTPaperViewOrCurrentView(directView)
      if let streamView = resolvedView as? VideoLiveStreamView {
        return streamView
      }
    }

    if let managerView = uiManager.view(forReactTag: viewTag) {
      let resolvedView = RCTPaperViewOrCurrentView(managerView)
      if let streamView = resolvedView as? VideoLiveStreamView {
        return streamView
      }
    }

    return nil
  }
}
